### ISSUE TITLE
Cannot subclass JsonReader and can't use a CharBuffer as JsonReader input

### ISSUE THREAD

**Issue by antoinebaudoux** [square/moshi#855]
Hello, 

I would like to deserialize from a list of `CharBuffer` objects (It could be any source of chars, like a single CharBuffer or a Reader, the problem would be the same). I don't want to re-decode the chars a second time.
I can't use any of the fromJson methods:

`fromJson(BufferedSource source)` is for byte sources. I already have decoded chars, I don't want to re-encode them only to re-decode them just after.
`fromJson(String string)` actually reencodes the string in utf8 only to re-decode it just after in JsonUtf8Reader. I don't want to do that.
`fromJson(JsonReader reader)` could work only if I could provide my own JsonReader implementation that takes `CharBuffer` objects as source. Unfortunately I can't subclass it.

It would be nice if you provided a JSONReader Implementation that takes a source of chars as input, in addition to JsonUtf8Reader that takes a source of bytes.
Or maybe allow subclassing JsonReader.

Thanks for your feedback,

Antoine

**Comment by NightlyNexus** [comment]
No need to encode the characters.
You could make an `okio.Source` implementation that reads from the CharBuffer, or I'd probably go ahead and write the CharBuffer to an `okio.Buffer`.
Then use JsonReader.of(BufferedSource).

**Comment by swankjesse** [comment]
Yeah, unfortunately Moshi is pretty optimized for reading UTF-8 sources, and when you write chars to a Buffer that’ll encode as UTF-8.

I’m converting this to a feature request to support reading from character streams directly. I don’t anticipate we’ll get to it _too_ quickly, because it’s not a case that’s typically a bottleneck.

**Comment by antoinebaudoux** [comment]
@swankjesse Thanks for your response, I guess I have no other choice for now than encode the chars in UTF-8
